oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/optimizer/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,151 @@
+    }
+
+    override fun stopOptimization(
+            request: StopOptimizationCommandDTO,
+            responseObserver: StreamObserver<StopOptimizationResponseDTO>
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStop(request)
+        }
+    }) { response ->
+        if (response.responseCase == StopOptimizationResponseDTO.ResponseCase.RUNID) apiService.stop()
+    }
+
+    override fun updateNode(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.updateNode(request)
+        }
+    }
+
+    override fun sendMessage(
+            request: MessageCommandDTO,
+            responseObserver: StreamObserver<MessageResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.sendMessage(request)
+        }
+    }
+
+    override fun unregisterRequest(
+            request: RequestUnRegistrationRequestDTO,
+            responseObserver: StreamObserver<UnRegistrationResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.unregister(request)
+        }
+    }
+
+    override fun offerSimulationResult(
+            request: SimulationResponseDTO,
+            responseObserver: StreamObserver<SimulationResultConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerResult(request)
+        }
+    }
+
+    override fun offerErrorResult(
+            request: ErrorResponseDTO,
+            responseObserver: StreamObserver<ErrorConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerError(request)
+        }
+    }
+
+    override fun offerSimulationConfig(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerConfig(request)
+        }
+    }
+
+    private fun <V> checkThenRun(hasPermission: Boolean, action: () -> V): V {
+        if (hasPermission) {
+            return action()
+        } else {
+            throw StatusRuntimeException(Status.PERMISSION_DENIED)
+        }
+    }
+
+    private suspend fun <V> checkThenRunAsync(hasPermission: Boolean, action: suspend () -> V): V {
+        if (hasPermission) {
+            return action()
+        } else {
+            throw StatusRuntimeException(Status.PERMISSION_DENIED)
+        }
+    }
```

Comment:
also, FYI, you dont need to duplicate this method if you tag the top one with `inline`. you can call `inline` functions from `suspend` funs. This is why things like `map` and `filter` work in `suspend` funs.

Reply:
ah, nice, I need to spend a bit more time reading inline